### ISSUE TITLE
Find-DbaInstance does not perform the SQLConnect test propertly

### ISSUE THREAD

**Issue by Shashtsoh** [dataplat/dbatools#8678]
### Verified issue does not already exist?

I have searched and found no existing issue

### What error did you receive?

None.  The command appears to always return true when the SQLConnect test is run, despite the test not being run at all.  Identified in issue #8676

### Steps to Reproduce

```powershell
find-dbainstance <computername> -scantype all | select MachineName, SQLInstance, Availability, SQLConnected
```

### Please confirm that you are running the most recent version of dbatools

1.1.143

### Other details or mentions

Refer to #8676 for more detail.  The issues are on Lines 504 and 527.

Line 504:
```powershell
$server = Connect-DbaInstance -SqlInstance $dataSet.FullSmoName -SqlCredential $SqlCredential
```
The property 'FullSmoName' does not exist, causing the connection attempt to always fail.

Line 527:
```powershell
if ($_.Exception.InnerException.Errors.Class -lt 25) {
                                    # There IS an SQL Instance and it listened to network traffic
                                    $dataSet.SqlConnected = $true
                                    $dataSet.Confidence = 'High'
                                }
```
When the connection fails, it hits this block in the catch statement.  The InnerException property is null due to the type of failure; the logic in the condition causes it to evaluate to true when this happens, leaving all connection attempts marked as successful with high confidence, despite the test not even being attempted.

### What PowerShell host was used when producing this error

PowerShell Core (pwsh.exe)

### PowerShell Host Version

Name                           Value
----                           -----
PSVersion                      7.2.7
PSEdition                      Core
GitCommitId                    7.2.7
OS                             Microsoft Windows 10.0.19045
Platform                       Win32NT
PSCompatibleVersions           {1.0, 2.0, 3.0, 4.0…}
PSRemotingProtocolVersion      2.3
SerializationVersion           1.1.0.1
WSManStackVersion              3.0

### SQL Server Edition and Build number

N/A - happens against all instances/versions

### .NET Framework Version

.NET 6.0.10

**Comment by Shashtsoh** [comment]
Replacing line 504 with below code has fixed the initial issue in my environment.  Connection attempts are made properly in all tests run so far.
```powershell
$server = if($dataset.BrowseReply) {
                                    Connect-DbaInstance -SqlInstance $dataSet.BrowseReply.SqlInstance -SqlCredential $SqlCredential
                                }
                                else {
                                    Connect-DbaInstance -SqlInstance $dataSet.SqlInstance -SqlCredential $SqlCredential
                                }
```

Replacing line 527 as follows fixes the null evaluation issue:
```powershell
if ($null -ne $_.Exception.InnerException -and $_.Exception.InnerException.Errors.Class -lt 25) {
```

However, there is still a potential issue with this second one.  One of my tests is against an machine with all SQL services stopped.  The connect attempt fails (as it should), but the error class in this scenario is 20, so we still evaluate to true and report that the connection was successful.  

Per the documentation for SqlException.Class:

> Messages that have a severity level of 10 or less are informational and indicate problems caused by mistakes in information that a user has entered. Severity levels from 11 through 16 are generated by the user, and can be corrected by the user. Severity levels from 17 through 25 indicate software or hardware errors. When a level 17, 18, or 19 error occurs, you can continue working, although you might not be able to execute a particular statement.

> The [SqlConnection](https://learn.microsoft.com/en-us/dotnet/api/microsoft.data.sqlclient.sqlconnection?view=sqlclient-dotnet-standard-5.0) remains open when the severity level is 19 or less. When the severity level is 20 or greater, the server ordinarily closes the [SqlConnection](https://learn.microsoft.com/en-us/dotnet/api/microsoft.data.sqlclient.sqlconnection?view=sqlclient-dotnet-standard-5.0). However, the user can reopen the connection and continue. In both cases, a [SqlException](https://learn.microsoft.com/en-us/dotnet/api/microsoft.data.sqlclient.sqlexception?view=sqlclient-dotnet-standard-5.0) is generated by the method executing the command.

Based on this, I'm not really sure what to do here.  Any threshold we set is going to have issues.  I think I'd recommend changing that condition to evaluate as true when the class is less 20 instead of 25.  I'm basing that off of the idea that the current condition sets 25 as the threshold because "There is a SQL instance, and it listened to traffic".  An offline instance cannot listen to traffic, and if we get a class 20 in that case, then the feels like the best option.  It may flag false when a network error results in a class 20, but it will never flag true when the instance is offline.

**Comment by Shashtsoh** [comment]
Thank @potatoqualitee!  I'm not really familiar with the way classes are defined in this project, but I'm happy to dig in on the code on my side if you can give me an idea of where to look.  The closest I've found are the xml definitions, but I haven't figured out how they tie in yet.  

I'm wondering if the reason that property doesn't yet exist at the time the test is attempted is because the instance of the instance parameter class we are working with is essentially an empty cast from the simple dnsname we provided to the computername parameter.  Because we're attempting to find instances, we don't yet know what instances are there, and therefore can't have detected the SMO name at that point.  I did find that FullSmoName does exist on the object returned by connect-dbainstance (who's main purpose appears to be to create an object compatible with the instance parameter), but that obviously doesn't exist in find-dbainstance until after the connection is attempted.

**Comment by Shashtsoh** [comment]
I think I found it - look like it's in DbaInstanceParameter.cs on lines 98-110.  It's been better than a decade since I've touched CS; I think I'll avoid PRs or code suggestions here.  That said, if I'm reading this correctly, this block assumes that the instance name is always going to include the computername that the function is fed.  For FCI's, this is still false.  
```cs
public string FullSmoName
        {
            get
            {
                string temp = _ComputerName;
                if (_NetworkProtocol == SqlConnectionProtocol.NP) { temp = "NP:" + temp; }
                if (_NetworkProtocol == SqlConnectionProtocol.TCP) { temp = "TCP:" + temp; }
                if (!String.IsNullOrEmpty(_InstanceName) && _Port > 0) { return String.Format(@"{0}\{1},{2}", temp, _InstanceName, _Port); }
                if (_Port > 0) { return temp + "," + _Port; }
                if (!String.IsNullOrEmpty(_InstanceName)) { return temp + "\\" + _InstanceName; }
                return temp;
            }
        }
```

**Comment by wsmelton** [comment]
Pending PR approval fix will be in 2.0 release.
